llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -4243,8 +4243,27 @@ SelectionDAG::computeOverflowForSignedAdd(SDValue N0, SDValue N1) const {
   if (ComputeNumSignBits(N0) > 1 && ComputeNumSignBits(N1) > 1)
     return OFK_Never;
 
-  // TODO: Add ConstantRange::signedAddMayOverflow handling.
-  return OFK_Sometime;
+  // smulhi + any value less than half of signed max
+  KnownBits N1Known = computeKnownBits(N1);
+  if (N0.getOpcode() == ISD::SMUL_LOHI && N0.getResNo() == 1) {
+    APInt Max = APInt::getSignedMaxValue(N1.getScalarValueSizeInBits());
+    Max.lshrInPlace(1);
+    if (N1Known.getMaxValue().sle(Max))
```

Comment:
If N1Known.getMaxValue() has bit 31 is will be considered negative number and it will be less than `Max` so you'll say it doesn't overflow. But if the result of the smulhi is negative, adding a negative number may overflow.

Reply:
Underflow actually, but is that the same thing in this context?

Reply:
For integers, LLVM uses "overflow" to refer to both directions.

Reply:
Do we have cases for SMUL_LOHI? The UMUL_LOHI equivalent patterns were pretty obscure and were to help fix a couple of specific UADDO+CARRY situations (see D29565\D29587).

You should be able to construct knownbits tests cases (although I expect many will already be caught by the signbits check above).

Reply:
Neither SMUL_LOHI or MULHS are supported in ComputeNumSignBits. Seems like we should start there before adding a special case to `computeOverflowForSignedAdd`.